### ISSUE TITLE
Question about speed of findAllElements

### ISSUE THREAD

**Issue by bubnenkoff** [renggli/dart-xml#113]
I need to find only first matched element. Now I am doing it:
```
final title =  document.findAllElements('title', namespace: '*').first.text;
```

But my XML docs are very big, and I suppose that `findAllElements` are doing full docs scanning. I have heard that some parsers can read only parts of docs, Can dart-xml do that?

**Comment by renggli** [comment]
Yes, check out the event-driven API: https://github.com/renggli/dart-xml#event-driven

**Comment by bubnenkoff** [comment]
@renggli  thanks!
Could you show an example how to use stream parsing to extract `title` value? And how to use namespaces `*`? 
```
parseEvents(bookshelfXml)
    .whereType<XmlTextEvent>()
    .map((event) => event.text.trim()) // what should be here? stmth like: event.tag == 'title'
    .where((text) => text.isNotEmpty)
    .forEach(print);
```

**Comment by bubnenkoff** [comment]
```
  final file = File(file_name);
  await file.openRead()
  .transform(utf8.decoder)
  .toXmlEvents()
  .selectSubtreeEvents((event) => event == 'ns6:title')
  .toXmlNodes()
  .forEach((node) => print(node));
  ```
  
  This code is from example.
  The first problem that node do not have method `innerText`
  The second problem -- it's do not clear how to ignore nasmaspeces. Could you help me please?

**Comment by renggli** [comment]
Regarding https://github.com/renggli/dart-xml/issues/113#issuecomment-964936451: This is not using a `Stream`, but a lazy `Iterable`. You need to keep track of the events and the nesting yourself. Typical strategy would be to use `Iterable. skipWhile` until the `XmlStartElementEvent` with the name `title` and then `Iterable.takeWhile` until the corresponding `XmlEndElementEvent`. From that sub-iterable you then extract the `XmlTextEvent` as in the example above.

Regarding https://github.com/renggli/dart-xml/issues/113#issuecomment-964960850: You seem to miss the import of the XML library, the `innerText` method is an extension method defined in the base package. For the matching remove the condition on the namespace prefix, you are comparing string fields. See [the code and class documentation](https://github.com/renggli/dart-xml/blob/main/lib/src/xml_events/utils/named.dart) for other accessors.

**Comment by bubnenkoff** [comment]
> innerText method is an extension method defined in the base package

But I have `import 'package:xml/xml.dart';` imported

**Comment by bubnenkoff** [comment]
It's seems that this code are work:
```
    final file = File(file_name);
    await file.openRead()
    .transform(utf8.decoder)
    .toXmlEvents()
    .selectSubtreeEvents((event) => event.localName == 'docPublishDate')
    .toXmlNodes()
    .forEach((node) => print(node[0].innerText));

```
Only question is `node[0]`. It's error in docs? But by idea node should not have any index.

**Comment by bubnenkoff** [comment]
In result I wrote next code.
The task sound like: to take any `datetime` from document:
```
  Future<String?> extractValue(List<String> variantList) async  {
    String? result;
      for(var el in variantList) {  

          final file = File(file_name);
          await file.openRead()
          .transform(utf8.decoder)
          .toXmlEvents()
          .selectSubtreeEvents((event) => event.localName == 'docPublishDate')
          .toXmlNodes()
          .forEach((node) {
             result = node[0].innerText;
          } );

          return result;

      }


    return result; 
  }

  print(await extractValue(["docPublishDate", "createDate", "plannedPublishDate", "docDT", "protocolDate", "publishDTInEIS", "publishDate"]));
```

It this code is ok? It's working. But maybe I should add any interruption/break? I need to find only first match. I do not need full scan of document.

**Comment by renggli** [comment]
You are right, before the `.forEach` there is a `.expand((nodes) => nodes)` missing in the tutorial.

**Comment by renggli** [comment]
For the rest have a look at Dart tutorials on Streams, finding the first element of a stream is a standard operation unrelated to this library.
